src/parser.ml

```diff
@@ -303,15 +303,17 @@ let parser arg_ident =
 
 (** Metavariable identifier (regular or escaped, prefixed with ['?']). *)
 let parser meta =
-  | "?" - id:{regular_ident | escaped_ident} -> in_pos _loc id
+  | "?" - id:{regular_ident | escaped_ident} ->
+      if id = "_" then Earley.give_up (); in_pos _loc id
```

Comment:
Perhaps we should exclude every ident starting with `_`. Perhaps we should introduce a new syntactic category for idents starting with `_`.

Reply:
I don't know, maybe. For now prefixing a (non-empty) identifier with `_` is only used as a way to disable the warning. Which I think is common in many languages.

Reply:
Note that, here, it is about metavariables. It could be interesting to write _xx to generate a fresh metavariable with a name prefixed by xx.

Reply:
I'm not sure I see the interest in that. You can already do `?abc[x,y,z]` to generate a metavariable named `abc`.

Reply:
But we will never do that. Imagine that you are in the middle of a proof with dozens of assumptions. You don't want to enumerate them all, but you may want to give an informative name to your metavariable. Moreover, currently, in proofs, goals, which are metavariables, are not referred to by their names but numbered instead (focus command), with a number that is not fixed. It could be useful at some point to be able to refer to some goal by its name. But this is not directly related to the current PR.

Meanwhile, and since _xx idents have a special meaning when used in an abstraction or a product, it would be good not to allow such idents for metas. So I propose to instead write `if id.[0] = "_" then Earley.give_up (); in_pos _loc id`.

Reply:
The thing is, they don't have a special meaning. Their only extra feature is to disable a warning. So I would just leave these things as they are no, until everything we want has been clarified. Remember that this is only a minor point of this PR, fixing a behavior that was not expected: we definitely do not want `?_` as a valid metavariable name. What we are talking about here has nothing to do with this PR. If you want to discuss that more, please open an issue and make a proposal for whatever semantics you think would be best.

Reply:
It's related. This PR is about wildcard's. I propose that any ident starting with _ is equivalent to _ alone and thus cannot be referred to later, and thus used, so that we could write neither \_x,_x nor &_x in a LHS and &_x in a RHS. Since metavariables can be refered to, they must not start with _.

Reply:
That is a big change in our lexical conventions, and I don't want to change that on a whim. This needs to be discussed more, and here is not the place. There is also another, potentially bigger issue with the LHS of rewriting rules: we disallow writing type annotations for λ-abstractions, but yet we allow constraining the involved types by forbidding certain λ-variables (those that are introduced with `_` since they are not added to the environment).

Reply:
What's the problem? And what's the relation with type annotations?

Reply:
The problem is that currently we argue that types should not be part of matching, which makes sense. So the user does not annotate the lambdas in rules, which I understand as "there is no constraint on the types". On the other hand, using `_` in a λ-abstraction introduces a constraint in the types of subsequent abstractions. There is an example of that in the test file I added (in the middle of the file). But roughly, if you have `λ x y, ...` in a pattern, then `x` can appear in the type of `y`. But if you have `λ _ y, ...` then the first bound variable cannot appear in the type of `y` (it is not put in the environment of the corresponding metavariable).

Reply:
Ok, I see. But I think that your comment: in tests/OK/292.lp is incorrect:
```
rule d2_1 (λx y, &_[y]) → λ_ _,zero // no warning
rule d2_1 (λ_ y, _    ) → λ_ _,zero // no warning
// The difference is that in the former variable "x" may appear in the type
// of the "y" variable. That is not the case in the second line.
```
From the point of view of matching, there should be no difference between the two rules: if a term is reducible by one rule, then it is reducible by the other. Changing the type of d2_1 to ∀x:R,Sx⇒T (or even ∀x:R,∀y:Sx,Txy) doesn't change any thing, as matching ignores types.
On the other, it may be important for checking subject-reduction. As there is no difference between these rules for the point of view of rewriting, when we translate pattern variables to metavariables, we need to take into account all variables, including the variables starting with underscore.
This means that we need to check that https://github.com/Deducteam/lambdapi/blob/ceba0a2383112aab85ec694be68ed4803316a27a/src/sr.ml#L72 is correct from this point of view, but this is completely orthogonal to this PR.

Reply:
Currently, a pattern variable of arity l is replaced by a metavariable of arity l (k=0 as pattern variables will never be applied after https://github.com/Deducteam/lambdapi/pull/279).

Reply:
Currently, if I understand well, `λx y, &_[y]` is equivalent to `λ(x:&A)(y:&B[x]),&C[y]` which in turn is translated to `λ(x:?A)(y:?B[x]),?C[y]`, while `(λ_ y, _    )` is equivalent to `λ(x:&A)(y:&B),&C[y]` which in turn is translated to `λ(x:?A)(y:?B),?C[y]`.

Do we need to fix to_m? An alternative is to fix https://github.com/Deducteam/lambdapi/blob/ceba0a2383112aab85ec694be68ed4803316a27a/src/scope.ml#L226 . Currently scope_domain generates a fresh pattern variable for domains (note by the way that we could put any dummy fixed term instead as domains are not matched). If in scoping, we keep track of all variables (those starting with underscore and those not starting with underscore) then, we can use the full environment in domains, and only the variables not starting with underscore else where.

So, again, it would be useful to replace string option by string, and filter out strings starting with underscore when necessary.

Reply:
> From the point of view of matching, there should be no difference between the two rules: if a term is reducible by one rule, then it is reducible by the other.

Yes, of course. But it leads to a difference when we check SR!

Reply:
> So, again, it would be useful to replace string option by string, and filter out strings starting with underscore when necessary.

I pretty much agree with all of what you said, but I think it is best to do these changes in a separate PR. The scoping module needs some refactoring and I'm going to get into conflicts with concurrent PRs if I go on now. Anyway, this PR already solves a couple of problems and it improves on the current state of the tool, so I don't see any reason not to merge it (after merging #288, and I rebase over it to cleanup the history).